controllers/fenceagentsremediation_controller.go

```diff
@@ -74,39 +75,70 @@ func (r *FenceAgentsRemediationReconciler) Reconcile(ctx context.Context, req ct
 	emptyResult := ctrl.Result{}
 
 	// Fetch the FenceAgentsRemediation instance
-	far := &v1alpha1.FenceAgentsRemediation{}
+	far := &farV1alpha1.FenceAgentsRemediation{}
 	if err := r.Get(ctx, req.NamespacedName, far); err != nil {
 		if apiErrors.IsNotFound(err) {
-			// FAR is deleted, stop reconciling
-			r.Log.Info("FAR CR is deleted - nothing to do", "CR Name", req.Name, "CR Namespace", req.Namespace)
+			// Request object of FAR CR was not found, could have been deleted after reconcile request.
+			// Return and don't requeue
+			r.Log.Info("FenceAgentsRemediation CR was not found", "CR Name", req.Name, "CR Namespace", req.Namespace)
 			return emptyResult, nil
 		}
-		r.Log.Error(err, "failed to get FAR CR")
+		r.Log.Info("Failed to get FenceAgentsRemediation CR, requeuing.", "CR Name", req.Name)
 		return emptyResult, err
 	}
+	// Add finalizer when object is created
+	if far.ObjectMeta.DeletionTimestamp.IsZero() {
+		controllerutil.AddFinalizer(far, farV1alpha1.FARFinalizer)
+		if err := r.Client.Update(context.Background(), far); err != nil {
+			return emptyResult, fmt.Errorf("failed to add finalizer to the CR - %w", err)
+		}
+	} else {
+		r.Log.Info("CR's deletion timestamp is not zero", "CR Name", req.Name)
+		// The object is being deleted
+
+		// remove node's taints
+		if err := utils.HandleTaint(r.Client, farV1alpha1.FARNoExecuteTaint, farV1alpha1.RemoveTaintAction, far); err != nil {
+			return emptyResult, err
+		}
+		// remove finalizer
+		controllerutil.RemoveFinalizer(far, farV1alpha1.FARFinalizer)
+		if err := r.Client.Update(context.Background(), far); err != nil {
+			return emptyResult, fmt.Errorf("failed to remove finalizer from CR - %w", err)
+		}
+		r.Log.Info("Finalizer was removed", "CR Name", req.Name)
+		return emptyResult, nil
+	}
+
 	// Validate FAR CR name to match a nodeName from the cluster
 	r.Log.Info("Check FAR CR's name")
 	valid, err := utils.IsNodeNameValid(r.Client, req.Name)
 	if err != nil {
 		return emptyResult, err
 	}
 	if !valid {
-		r.Log.Info("didn't find a node matching the CR's name", "CR's Name", req.Name)
+		r.Log.Info("Didn't find a node matching the CR's name", "CR's Name", req.Name)
```

Comment:
IMO this scenario might need something more visible than Info log

Reply:
Such as? Something in UI?

Reply:
At least an error log, maybe an event.